30_regexp_lab/relab.txt

```diff
@@ -1,17 +1,7 @@
-Use grep to print only the lines from "input" for which:
-
-1. There is a binary number in the line
-
-2. There is an even binary number in the line
-
-3. Line includes a hexadecimal number
-
-4. Line starts with a lowercase letter
-
-5. Line is not empty
-
-6. Line starts with a lowercase letter and ends in a number
-
-7. Line is longer than 50 characters
-
-8. Line is of an even length
+1. grep -E '[0-9]{0,1}' input | head -11
+2. 
+3. grep hexadecimal input
+4. grep ^[a-z] input
+5. grep -v -e '^$' input
+6. grep '[^a-z] [$0-9]' input

```

Comment:
1. In Regular expressions a space does not mean AND but just a space
2. Inside [...] the signs ^ and $ lose their meaning

so your expression translated to words is: "all lines that contain one character that is not a lowercase letter, then a space, and then a single character that is either a $ or a digit."

A correct version looks like this: `^[a-z].*[0-9]$`. Here's why:

1. `^` is the beginning of a line
2. `[a-z]` next comes a lowercase letter
3. `.*` next comes a sequence of characters. `.` means any character and the following `*` means that thing can appear any number of times.
4. `[0-9]` then a single digit
5. `$` and finally the end of the line. Nothing should ever appear after a `$` because no characters can appear after the end of the line